### PR TITLE
Submit Int6 Native + SmearGate + BigramHash config

### PR DESCRIPTION
**Pull request by metadore**

### CODE REVIEW

**Comment by MatoTeziTanka** [comment]
### Community Review — Submit Int6 Native + SmearGate + BigramHash config

**Compliance:** LOOKS CLEAN — pure-neural submission, no TTT/SLOT/n-gram-cache

## PR #1373 — `int6_smeargate_bigramhash` Audit **Head SHA:** 83364e82e2340dc7cb32e2f1bea21c63f41ff88a **Files changed:** records/int6_smeargate_bigramhash/{README.md, submission.json, train_gpt.py} --- ### BigramHash check (ILLEGAL n-gram family bug) Lines 725–728 implement the bigram hash: ```python prev_ids = torch.cat([torch.zeros_like(input_ids[:, :1]), input_ids[:, :-1]], dim=1) bigrams = (prev_ids * self.tok_emb.num_embeddings + input_ids) % self.bigram_hash_size x = x + self.bigram_hash(bigrams) ``` The hash key is `(prev_id * vocab_size + cur_id) % bigram_hash_size`. This is a standard Cantor/linear-index pairing — **target token (`target_ids`) is NOT XOR'd or mixed into the key**. `target_ids` is only used in the loss computation at line 757. The hash key is derived solely from `input_ids` (past context), which is legal. No ILLEGAL n-gram family bug. The `smear_gate` (lines 730–733) is a learned gating on the previous embedding, also derived only from past tokens. Legal. --- ### TTT check (Pre-Quant / score-first / val_tokens) No test-time training of any kind is present. There is no loop over `val_tokens` that computes gradients, no optimizer step on validation data, no `is_last_chunk` guard, no scored-region slot logic. The only use of `val_tokens` is read-only inference inside `eval_val` (lines 255–284, under `torch.inference_mode()`). No TTT pattern of any kind. --- ### Post-quant roundtrip Lines 1112–1155: standard int8+zlib quantize → save → reload → re-evaluate. No parameters updated. Legal. --- ### Summary This is a clean pure-neural submission. Architecture additions are: (1) BigramHash embedding keyed on legal (prev_id, cur_id) pairs from input context only; (2) a learnable smear gate over previous token embeddings; (3) U-Net skip connections with `skip_weights`. All three are trained...

**Verdict:** LOOKS CLEAN.

**Recommendation to @cocohearts @valerio-oai @0hq @yuzhougu-oai @notapplica:** **MERGE** pending the usual record-track checks (3-seed validation, under-16MB artifact cap, ≤600s train + ≤600s eval on 8×H100 SXM). No compliance flags from the audit — this looks like a clean pure-neural submission.

---
*Reviewed by [@MatoTeziTanka](https://github.com/MatoTeziTanka) — [The Agora](https://matotezitanka.github.io/parameter-golf). Compliance audit via LLM agent (Sonnet) reviewing full train_gpt.py source, cross-checked against deterministic AST classifier. If this review misread your code, please call it out so I can re-audit manually.*
